### DISCUSSION TITLE
3주차 사전 과제, 김경진

### DISCUSSION THREAD

**Discussion by lessmadmax** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#27] [General]
1. 
- 소셜 미디어 앱을 개발 중인 상황에서, 사용자의 프로필 정보, 게시물 목록, 친구 목록 등 다양한 정보를 관리해야 합니다.
- 이 상황에서 다양한 종류의 정보를 어떻게 구조화하고 관리할 건가요? 여러분이 생각한 방식과 그 이유를 설명해주세요. 구체적인 기술이나 라이브러리 이름을 제시하지 않아도 좋습니다.

데이터베이스를 사용할 것입니다. 

2. 
<img width="335" alt="스크린샷_2024-07-21_오후_9 35 22" src="https://github.com/user-attachments/assets/ac433508-bf54-415c-ad99-8311da3f26ea">

아래 코드는 위와 같은 간단한 사용자 프로필 관리 애플리케이션 코드의 일부입니다. 

```
// App.jsx
function App() {
  const [user, setUser] = useState({
    name: 'John Doe',
    theme: 'light'
  });

  return (
    <div>
			<header>Welcome, {user.name}</header>;
      <Main user={user} setUser={setUser} />
    </div>
  );
}
```
```
// Main.jsx
function Main({ user, setUser }) {
  return (
    <main>
	    <div>
	      <h2>Profile</h2>
	      <ThemeSwitcher user={user} setUser={setUser} />
	    </div>
	    <div style={{background: user.theme === 'light' ? '#fff' : '#333', color: user.theme === 'light' ? '#000' : '#fff'}}>
	      <h2>Content</h2>
	      <p>This content respects user's theme preference.</p>
	    </div>
    </main>
  );
}
```
```
//ThemeSwitcher.jsx
function ThemeSwitcher({ user, setUser }) {
  const toggleTheme = () => {
    setUser({...user, theme: user.theme === 'light' ? 'dark' : 'light'});
  };

  return (
    <button onClick={toggleTheme}>
      Switch to {user.theme === 'light' ? 'Dark' : 'Light'} Theme
    </button>
  );
}
```
- 제시된 코드에서 `ThemeSwitcher` 컴포넌트에서 테마를 변경하면, 어떤 과정을 거쳐 `Main` 컴포넌트의 스타일이 업데이트 되는지 `user` 데이터의 전달 흐름을 적어주세요.

(1) 'ThemeSwitcher' 컴포넌트에서 버튼을 클릭하면 toggleTheme 함수가 호출 됨
(2) toggleTheme 함수는 user 객체에 theme이라는 속성을 추가하고 해당 속성의 값을 조건에 따라 light 또는 dark로 결정함
(3) Main 컴포넌트의 div 태그의 style은 user 객체의 theme 속성에 따라 흰색 배경, 검은색 배경으로 바뀜

- 위 문제의 답변을 작성하는 과정에서 코드를 읽었을 때 들었던 코드에 대한 감상을 자유롭게 적어주세요.

충분히 가독성있고 데이터의 흐름을 파악하기 쉽다고 생각했습니다.

3.
<img width="290" alt="스크린샷_2024-07-21_오후_9 50 36" src="https://github.com/user-attachments/assets/c48fbf73-5036-4d02-83e3-741280c5ecad">

아래 코드는 위와 같은 UI의 온라인 쇼핑몰 장바구니 관리 코드입니다. 현재 장바구니에 상품을 추가, 제거하는 동작이 있습니다.

```
//App.jsx
function App() {
  const [cart, setCart] = useState([]);

  const addToCart = (item) => {
    setCart([...cart, item]);
  };

  const removeFromCart = (itemId) => {
    setCart(cart.filter(item => item.id !== itemId));
  
  return (
    <div>
      <ProductList addToCart={addToCart} />
      <Cart 
        items={cart} 
        removeFromCart={removeFromCart} 
      />
    </div>
  );
}
```
```
//ProducList.jsx
function ProductList({ addToCart }) {
  const products = [
    { id: 1, name: "상품 1", price: 1000 },
    { id: 2, name: "상품 2", price: 2000 },
  ];

  return (
    <div>
      {products.map(product => (
        <div key={product.id}>
          <h3>{product.name}</h3>
          <p>가격: {product.price}원</p>
          <button onClick={() => addToCart(product)}>장바구니에 추가</button>
        </div>
      ))}
    </div>
  );
}
```
```
//Cart.jsx
function Cart({ items, removeFromCa }) {
  const total = items.reduce((sum, item) => sum + item.price, 0);

  return (
    <div>
      <h2>장바구니</h2>
      <h3>장바구니 상품 개수: {cart.length}</h3>
      {items.map(item => (
        <div key={item.id}>
          <span>{item.name} - {item.price}원</span>
          <button onClick={() => removeFromCart(item.id)}>제거</button>
        </div>
      ))}
      <p>총 가격: {total}원</p>
    </div>
  );
}
```
- 지난 3주차 까지의 스터디 내용을 바탕으로 만약 상품의 수량을 변경하는 기능을 추가하려면 어떻게 구현할지 그 방법을 적어주세요.

상품의 수량을 변경하려면 event handling을 통하여 기능을 구현할 수 있습니다. 리스트 products를 useState를 통하여 초기화하고 count 속성을 추가합니다. 수량 조절 버튼을 추가하여 해당 버튼을 클릭하면 count의 값을 변경시키는 event로 상품의 수량을 변경하는 기능을 추가할 수 있습니다.

**Comment by MrMirror21** [comment]
작성해주신 사전과제 답변 잘 봤습니다. 
1번의 경우 프론트 단에서 데이터를 다루는 방법에 대해서 말씀드린건데 제가 조건을 잘못 적은 것 같네요. 일반적으로는 백 단의 데이터베이스에서 관리하는게 맞으나, 회원가입 과정이나 프로필 페이지를 보여주는 과정 등 여러 페이지에 걸쳐서 사용자의 개인 정보를 다뤄야 하는 순간들이 있습니다. 그런 점에서 오늘 스터디에서 말씀드린 State, Context 나 전역 변수 라이브러리를 사용하시면 될 것 같습니다. 
2,3번의 경우 훌륭하게 잘 대답해주셨네요. 해당 사전과제가 props drilling과 useContext, useReducer를 다뤘던 스터디 내용 학습하시는데 도움이 됐으면 합니다.
